[Crash fix] Fix cudaMallocAsync crashes.

nouiz opened the pull request:
The first commit fixes #48869. The second commit fixes another follow up crashes when using TF_GPU_ALLOCATOR=cuda_malloc_async.

The 2 fixes are:
- The Allocator API have the statistics optional. But the function [BaseGPUDeviceFactory::CreateGPUDevice()](https://github.com/tensorflow/tensorflow/blob/aa08697dcb98135ee39ba00ee08b5c1a28cfde61/tensorflow/core/common_runtime/gpu/gpu_device.cc#L1403) check that they are available.
- Bad handling of the ptr to the compute stream. It is a ptr to the compute stream that is passed, not the compute stream itself.
@sanjoy @chsigg

nouiz committed:
[Crash fix] Handle correctly the passed stream.

nouiz committed:
[Crash fix] Other part of TF need the stats for GPU allocator.

nouiz committed:
Add a cudaMallocAsync test

sanjoy commented on tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.h:
```diff
@@ -85,7 +85,7 @@ class GpuCudaMallocAsyncAllocator : public Allocator {
   void SetStream(void* stream) override {
 #if TF_CUDA_MALLOC_ASYNC_SUPPORTED
-    cuda_stream_ = reinterpret_cast<CUstream>(stream);
+    cuda_stream_ = *(reinterpret_cast<CUstream*>(stream));
```
I did not notice this earlier, but a setter for an immutable property seems like a code smell to me.  Have you tried passing in the stream via the constructor somehow?

(Does not need to be addressed in this PR.)

nouiz replied:
Mostly, the GpuDevice need the gpu_allocator. But the gpu_allocator need the stream created by the GpuDevice.
Fixing this would request a refactoring that will break existing API and I wasn't sure if those API was part of the public API or not. As I wanted this to get in TF2.5, I prefered to stay away from this refactoring.
Relevant code:
https://github.com/tensorflow/tensorflow/blob/aa08697dcb98135ee39ba00ee08b5c1a28cfde61/tensorflow/core/common_runtime/gpu/gpu_device.cc#L1414-L1422

(I also updated the PR description)
(I amended the test commit to remove a commented left over line)

sanjoy reviewed (changes requested):
Can you please update the PR description to state what the bug was and how this PR fixes it?

nouiz committed:
Use a simpler static_cast

nouiz committed:
Make the new options experimental and add it to the golden list of public API.

nouiz committed:
Use the new option

nouiz committed:
Test that the new allocator is instantiated, not that there is no crash.

nouiz commented:
I updated this PR. Now the new configuration options is experimental.
I added what got lost in my rebase the use of that new configuration option.
I also updated the test to verify that the right allocator got used. If you have another idea how to do this, tell me.

sanjoy commented on tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.h:
```diff
@@ -106,6 +110,10 @@ class GpuCudaMallocAsyncAllocator : public Allocator {
   CUmemoryPool pool_;
 #endif  // TF_CUDA_MALLOC_ASYNC_SUPPORTED
 
+  // Just a counter for the number of time this class is instanciated.
```
s/instanciated/instantiated/ everywhere.

Also s/nb/number

nouiz replied:
Done.

sanjoy commented on tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.h:
```diff
@@ -106,6 +110,10 @@ class GpuCudaMallocAsyncAllocator : public Allocator {
 
+  // Just a counter for the number of time this class is instanciated.
+  // Only useful for tests.
+  static int nb_instanciated_;
```
Does this need to be an `std::atomic`?

nouiz replied:
Implemented. Note the current TF GpuDevice create them serially. But having it atomic is more future proof.

sanjoy commented on tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.h:
```diff
@@ -85,10 +85,14 @@ class GpuCudaMallocAsyncAllocator : public Allocator {
 #endif
   }
 
+  static int NbInstantiated() {
```
Maybe call this `GetInstantiatedCountTestOnly`?

nouiz replied:
Done.

sanjoy reviewed (changes requested):
Minor comments.

nouiz committed:
Variable/function renaming. Also make a counter atomic to be more future proof. The current initialization code in TF instantiate then serially.

nouiz commented:
Just checking if things are good on your side. It was approved 6 days ago and didn't get in yet.

penpornk commented:
@nouiz Could you please help resolve the merge conflict? Thank you!

nouiz commented:
Rebased.

penpornk commented:
@nouiz Thank you! It seems there are still conflicting files though, namely:
```
tensorflow/core/protobuf/config.proto
tensorflow/tools/api/golden/v1/tensorflow.-g-p-u-options.pbtxt
```
Could you please help resolve the conflicts again? Sorry for the inconvenience!

nouiz committed:
Fix a segfaul introduced in 76e61092a7aa8958ea190453d1c51a119de3e189. sub_allocator is owned by gpu_bfc_allocator. deleting it early doesn't any real benefit. So lets the normal clean up do its work.

nouiz commented:
I fixed the new conflict.
I also found a bug in the changes that caused the first merge conflict. I added the fix as it made my tests fail in CUDA 11.3.

penpornk commented on tensorflow/core/common_runtime/gpu/gpu_device_test.cc:
```diff
@@ -114,6 +117,33 @@ class GPUDeviceTest : public ::testing::Test {
+    EXPECT_EQ(devices.size(), 1);
+    Device* device = devices[0].get();
+    auto* device_info = device->tensorflow_gpu_device_info();
+    CHECK(device_info);
```
Could you please use `DCHECK` instead? We don't use `CHECK` in new code anymore. Sorry we didn't notice this earlier!

nouiz replied:
This is a test. I really want a crash even in no-debug mode.
So I pushed a change to use an EXPECT instead.

penpornk reviewed (changes requested):
Thank you very much!

nouiz committed:
Replace a CHECK to EXPECT_NE.

ruler501 commented on tensorflow/core/protobuf/config.proto:
```diff
@@ -208,6 +208,9 @@ message GPUOptions {
+
+    // When true, use CUDA cudaMallocAsync API instead of TF gpu allocator.
+    bool use_cuda_malloc_async = 11;
   }
```
Is this correct? Assigning an integer constant to a bool seems wrong.

akuegel replied:
This is not assigning a value, but assigning an ID. So this is fine.

akuegel commented:
> Just to be sure, I see many "Internal CI build failed". Do they block this PR? What is their cause? The windows error is that bazel didn't found the strategy "sandboxed". That have been reverted to my knowledge.

The problem was an unused variable. @penpornk has adjusted your patch internally, then that patch needed approval internally, now it is theoretically ready to land (if the tests pass).

nouiz commented:
Thanks for the update.

nouiz commented:
In some of the refactoring of cudaMallocAsync, when there is an OOM a deadlock was added.
Here is a commit that fix it: https://github.com/tensorflow/tensorflow/commit/4cab1820acd4f9fa031bc59884e4bc461d33c3fa

I wasn't sure if you wanted to include it in this PR or not. I can make another PR with the trivial fix if you want. Just tell me how you want to handle it.

akuegel commented:
> In some of the refactoring of cudaMallocAsync, when there is an OOM a deadlock was added.
> Here is a commit that fix it: [4cab182](https://github.com/tensorflow/tensorflow/commit/4cab1820acd4f9fa031bc59884e4bc461d33c3fa)
> 
> I wasn't sure if you wanted to include it in this PR or not. I can make another PR with the trivial fix if you want. Just tell me how you want to handle it.

I think landing that separately is fine. As far as I can tell, the problem is a missing BUILD dependency for the new header include in the gpu_device_test. I added this as review comment on the internal change, hopefully that is the only remaining thing needed to fix to get it landed.

nouiz commented:
Any update? I think TF2.6 cutoff is soon. It would be great to get it included so people can start to use it manually.
Is there anything I can do to help?

penpornk commented:
@nouiz Apologies for the delay! A couple of tests failed the heapcheck (memory leaks). (You'll have to build them and run the binary with pprof to reproduce. The problem is these tests seem to have problems building in my open source environment right now -- so I'm not sure if you'll be able to build them.)
```
$ bazel build --config=cuda --config=opt  \
   //tensorflow/core/common_runtime/gpu:gpu_device_test
$ bazel build --config=cuda --config=opt \
   //tensorflow/core/common_runtime/gpu:gpu_device_test_gpu
```

Any idea where the leak could come from? I'm sorry I haven't had time to take a closer look at what the PR is doing yet. I'll try to do that today. Will try to get this into 2.6.

nouiz commented:
I looked at the code. But I do not understand what could leak.
Here there is a different cast: https://github.com/tensorflow/tensorflow/pull/49173/files#diff-d54c8000677cfd5b7bbc09963559b66949e9738b944210ee215dff19730ee296R88

I do not know pprof. Is there any options I must specify? I'll try to reproduce the error. Just in case I can help.

penpornk commented:
@nouiz Thank you very much for offering to help! :)
I think I've fixed the leaks. (I added the code that delete `gpu_bfc_allocator` back. But not the suballocator.) 
```c++
      delete gpu_bfc_allocator;
      gpu_bfc_allocator = nullptr;
```

The merge is now stuck because of some (seemingly unrelated) infra issues instead. I'll keep you posted.

Just for reference, here is how to use `pprof` to check for memory leaks (paths are based on Ubuntu 18.04 -- may be different on different OSes):
```
# Install pprof and TCMalloc
$ sudo apt install libgoogle-perftools-dev google-perftools
$ export PPROF_PATH=/usr/bin/google-pprof

# Build the test
$ bazel build --config=cuda --config=opt //tensorflow/core/common_runtime/gpu:gpu_device_test

# Run the test with TCMalloc and pprof's heapcheck enabled.
$ HEAPCHECK=normal LD_PRELOAD=/usr/lib/libtcmalloc.so TF2_BEHAVIOR=1 \
    bazel-bin/tensorflow/core/common_runtime/gpu/gpu_device_test
```

penpornk commented:
Thank you very much again for your patience!

> I do not see a call to pprof. Shoudn't it be:
> ```
> HEAPCHECK=normal LD_PRELOAD=/usr/lib/libtcmalloc.so TF2_BEHAVIOR=1
> $PPROF_PATH bazel-bin/tensorflow/core/common_runtime/gpu/gpu_device_test
> ```
> Or something like that?

Great question. [Heap checker](https://gperftools.github.io/gperftools/heap_checker.html) is part of TCMalloc. So we only need to link the binary to TCMalloc and turn on heap checker with the environment variable `HEAPCHECK`. It uses `pprof` underneath, that's why we need to install and set the path to `pprof`. :)
